File: Python/0240_Search_A_2D_Matrix_II.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def search(row_idx):
            left, right = 0, len(matrix[row_idx])
            while left < right:
                mid = left + right >> 1
                if matrix[row_idx][mid] >= target:
                    right = mid
                else:
                    left = mid + 1
            return left
                
        for row_idx in range(len(matrix)):
            idx = search(row_idx)
            if idx < len(matrix[row_idx]) and matrix[row_idx][idx] == target:
                return True
        return False

File: CPlusPlus/0240_Search_A_2D_Matrix_II.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target){
        int lhs = 0, rhs = nums.size();
        while (lhs < rhs) {
            int mid = lhs + rhs >> 1;
            if (nums[mid] >= target) {
                rhs = mid;
            } else {
                lhs = mid + 1;
            }
        }
        
        return lhs;
    }
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size(), cols = matrix[0].size();
        for (auto& row : matrix) {
            int idx = search(row, target);
            if (idx < cols && row[idx] == target) {
                return true;
            }
        }
        return false;
    }
};
